refactor: Add generated docs, change option names

levouh opened the pull request.

levouh committed:
doc: Generate vimdocs, change option names

levouh committed:
refactor: Change option names

- Option name changes are backwards compatible
- Move color-related code into its own module
- Add in-line function documentation
- Change local/function parameter names to make more sense

levouh commented:
@williamboman adding docs and refactoring some stuff. Would love a review if you have a minute, otherwise no worries.

williamboman commented on DOC.md:
```diff
@@ -0,0 +1,60 @@
+*type*: `function`
+*default*: `nil`
+
+A function that will be called for each window to discern whether or not it should be tinted. Arguments are are `(winid)`, return `false` or `nil` to tint a window, anything else to not tint it.
```
```suggestion
A function that will be called for each window to discern whether or not it should be tinted. Arguments are `(winid)`, return `false` or `nil` to tint a window, anything else to not tint it.
```

williamboman reviewed (approved):
Looks good! I was thinking maybe the color transformation functions could be more composable, to better control the order of transformations, make it more extensible, and allow custom transformations. Would probably help with testability too! I'm thinking something along the lines of

```lua
colors.saturate = function (saturation)
    return function (r, g, b)
        local rec601_luma = 0.299 * r + 0.587 * g + 0.114 * b

        r = math.floor(r * saturation + rec601_luma * (1 - saturation))
        g = math.floor(g * saturation + rec601_luma * (1 - saturation))
        b = math.floor(b * saturation + rec601_luma * (1 - saturation))
        return r, g, b
    end
end

colors.tint = function (tint)
    return function (r, g, b)
        return r + tint, g + tint, b + tint
    end
end

colors.transform_color = function (hex, transformers)
    local r, g, b = colors.hex_to_rgb(hex)
    for _, transformer in ipairs(transformers) do
        r, g, b = transformer(r, g, b)
    end
    return colors.rgb_to_hex(r, g, b)
end
```

and then to call it

```lua
colors.transform_color("#8493B1", {
    colors.saturate(0.7),
    colors.tint(-20),
})
```

edit: This could also potentially be reflected in the configuration API, where instead of taking a "tint" and "saturation" amount, it could simply accept a table of transformations, with the default being:

```lua
require("tint").setup {
    transforms = {
         require("tint.colors").saturate(0.7),
         require("tint.colors").tint(-40),
    }
}
```

levouh committed:
refactor: compose color modification functions

levouh committed:
doc: Generate docs

levouh committed:
chore: Update in-line docs for new autocommands

levouh committed:
feat: Expose transforms directly to the user

levouh commented:
A handful of changes (last 4 commits) if you have time to review again @williamboman. If not, zero worries.

Thanks for the suggestions.

williamboman commented on lua/tint/colors.lua:
```diff
@@ -0,0 +1,83 @@
+---@param hex string The hex color to transform.
+---@param transforms table A table of functions used to transform the input hex color
+---@return string The hex representation color transformed by the configured values
+colors.transform_color = function(hl_group_info, hex, transforms)
```
Seems like `hl_group_info` is missing a `@param` definition (pointing this out because it's not immediately clear to me what it is, so I was looking for its explanation)

levouh replied:
Fixed. Let me know if that adds enough clarity.

williamboman commented on lua/tint.lua:
```diff
@@ -1,121 +1,142 @@
+local __ = {
+  transforms = {
+    [tint.transforms.SATURATE_TINT] = {
+      colors.saturate(tint.config.saturation),
+      colors.tint(tint.config.tint),
```
Wouldn't this need to be evaluated after the `.setup()` function has been called. By just looking at this it seems like it'll only ever use the default values, but I might be missing something?

levouh replied:
Yes, I just changed it to be a function that gets called in `get_transforms` to fix this issue when testing it locally.

levouh replied:
The way I fixed it was wrong actually - should be fixed for real now.

levouh committed:
fix: Handle changes to colorscheme

- User configuration only set once, used properly multiple times
- Move more things under private namespace
- Default transforms as function to pull user configuration values
- Fix in-line docs

williamboman reviewed (commented):
Looks great! I wonder if the `SATURATE_TINT` constant could be avoided so that the type of the `transforms` config could be fixated to a single type: an _optional_ table of transformations? If it's `nil` (the default) the default transforms would apply `{ colors.saturate(tint.config.saturation), colors.tint(tint.config.tint) }`. This could either take place in the `.setup()` function, or when the config item is read (`config.transforms or { ... }`).

I think having a predefined set of transforms makes a lot of sense, but maybe it's best to wait until there's a well-defined use case for it?

This is just nitpicking at this point though, so don't let me get in the way haha. I only really have one comment regarding how (or rather when) the default `SATURATE_TINT` is constructed

levouh commented:
>I wonder if the SATURATE_TINT constant could be avoided so that the type of the transforms config could be fixated to a single type: an optional table of transformations?

I think this is completely valid. My fear is that like what you did with the addition of saturation, people would end up adding useful transformation but then only keeping them locally. I'm not sure this really "solves" that problem, but it seems like it _might_ encourage people to expose them via a PR here if there is already a predefined way to do so.

Thoughts?

levouh commented on lua/tint.lua:
```diff
@@ -1,121 +1,147 @@
+---@param hl_group_name string
+---@param hl_def table The highlight definition, see `:h nvim_set_hl`
+local function set_tint_ns(hl_group_name, hl_def)
+  local hl_group_info = { hl_group_name = hl_group_name }
```
_fyi_: `hl_group_info` defined here - basically just a way for users in their various `transform` functions to do things "dynamically" based of of what color they are transforming. E.g. if they wanted to handle one color differently than another.

levouh commented on lua/tint.lua:
```diff
@@ -1,121 +1,147 @@
+--- Get the set of transforms to apply to highlight groups from the colorscheme in question
+---
+---@return table A table of functions to transform the input RGB color values by
+local function get_transforms()
```
This whole function might be something easier done once when the configuration is setup rather than each time `setup_all` is called (which only really happens once given the colorscheme is not changed).

Not the end of the world, but could be made a bit simpler.

williamboman commented:
> I think this is completely valid. My fear is that like what you did with the addition of saturation, people would end up adding useful transformation but then only keeping them locally. I'm not sure this really "solves" that problem, but it seems like it _might_ encourage people to expose them via a PR here if there is already a predefined way to do so.
> 
> Thoughts?

Yeah I've had this issue with my plugins. For mason.nvim I've added a [CONTRIBUTING.md](https://github.com/williamboman/mason.nvim/blob/main/CONTRIBUTING.md) and a more in-depth [API reference](https://github.com/williamboman/mason.nvim/blob/main/doc/reference.md), but I've not really been able to measure its impact. One thing I've observed is that even if you've designed something to not be extensible from the outside, people will still find a way and do it. So making something intrinsically extensible will probably cause a lot of private customization that people don't end up sharing/contributing.

Maybe a mention somewhere in the README as well as an issue/discussion where people can share their configs (screenshot + code) could help?

levouh committed:
fix: Ensure `VimEnter` has happened before loading

levouh committed:
doc: Add `CONTRIBUTING` doc to hopefully promote upstreaming useful tranforms

levouh committed:
chore: Generate docs

levouh committed:
fix: Call function in both cases when transform value is a string

levouh committed:
refactor: Only call `get_transforms` once

levouh committed:
chore: Messages when `vim.validate` fails

levouh commented:
This is completely my fault - I've ended pushing way too much stuff to this branch that brings it out of scope of what I'd originally made it for.

Going to merge it as-is for now, but happy to iterate on things through other (smaller + focused :wink:) PRs. Thank you for the feedback @williamboman.

williamboman commented:
Haha it happens. I hope I was of any help, great work!
